### PR TITLE
KAFKA-14752: Kafka examples improvements - demo changes

### PR DESCRIPTION
**Pull request by fvaleri**
This is extracted from the original PR for better review. https://github.com/apache/kafka/pull/13492

### CODE REVIEW

**Comment by showuon** [comment]
Please resolve the conflict. Thanks.

File: examples/src/main/java/kafka/examples/KafkaConsumerProducerDemo.java [hunk 1]
```diff
@@ -16,29 +16,59 @@
...
+
+            // stage 3: consume records from topic1
+            Consumer consumerThread = new Consumer(
+                "consumer", BOOTSTRAP_SERVERS, TOPIC_NAME, GROUP_NAME, Optional.empty(), false, numRecords, latch);
+            consumerThread.start();
+
+            if (!latch.await(5, TimeUnit.MINUTES)) {
+                Utils.printErr("Timeout after 5 minutes waiting for termination");
+                //producerThread.shutdown();
+                consumerThread.shutdown();
```
Comment (showuon) [inline examples/src/main/java/kafka/examples/KafkaConsumerProducerDemo.java]: Why does it get commented out?

File: examples/src/main/java/kafka/examples/KafkaExactlyOnceDemo.java [hunk 2]
```diff
@@ -16,182 +16,90 @@
...
- *    the output topic, while the message ordering on partition level is maintained.
- *    The driver will block for the consumption of all committed records.
+ * 1. Clean any topics left from previous runs.
+ * 2. Set up a producer thread to pre-populate a set of records with even number keys into the input topic.
+ *    The demo will block for the record generation to finish, so the producer is synchronous.
+ * 3. Set up the transactional instances in separate threads, each one executing a read-process-write loop
+ *    (See {@link ExactlyOnceMessageProcessor}). Each EOS instance will drain all records from either given
+ *    partitions or auto assigned partitions by actively comparing log end offset with committed offset.
+ *    Each record will be processed exactly once, dividing the key by 2 and extending the value record.
+ *    The demo will block until all records are processed and written to the output topic.
```
Comment (showuon) [inline examples/src/main/java/kafka/examples/KafkaExactlyOnceDemo.java:36]: The demo will block until all records are processed and written to the output topic[, with transactional guarantee.]

Adding the last one like before, to emphasize it.

File: examples/src/main/java/kafka/examples/KafkaExactlyOnceDemo.java [hunk 3]
```diff
@@ -16,182 +16,90 @@
...
+ * 4. Create a read_committed consumer  thread to verify we have all records in the output topic,
+ *    and record ordering at the partition level is maintained.
+ *    The demo will block for the consumption of all committed records.
  *
- * From this demo, you could see that all the records from pre-population are processed exactly once,
- * with strong partition level ordering guarantee.
```
Comment (showuon) [inline examples/src/main/java/kafka/examples/KafkaExactlyOnceDemo.java:69]: Could we keep this sentence?

File: examples/src/main/java/kafka/examples/KafkaExactlyOnceDemo.java [hunk 4]
```diff
@@ -16,182 +16,90 @@
...
+            // stage 3: read from input-topic, process once and write to the output-topic
+            CountDownLatch processorsLatch = new CountDownLatch(numInstances);
+            for (int instanceIdx = 0; instanceIdx < numInstances; instanceIdx++) {
+                ExactlyOnceMessageProcessor messageProcessor = new ExactlyOnceMessageProcessor(
+                    INPUT_TOPIC, OUTPUT_TOPIC, instanceIdx, processorsLatch);
+                messageProcessor.start();
+            }
+            if (!processorsLatch.await(2, TimeUnit.MINUTES)) {
+                Utils.printErr("Timeout after 2 minutes waiting for record copy");
+                //processors.forEach(ExactlyOnceMessageProcessor::shutdown);
```
Comment (showuon) [inline examples/src/main/java/kafka/examples/KafkaExactlyOnceDemo.java]: why commented out?

**Review by showuon — COMMENTED** [review]
Left some comments, and remember to resolve conflict. Thanks.

File: examples/src/main/java/kafka/examples/KafkaConsumerProducerDemo.java (hunk 1 above)
Comment (fvaleri) [reply examples/src/main/java/kafka/examples/KafkaConsumerProducerDemo.java]: Rebase issue, fixed.

File: examples/src/main/java/kafka/examples/KafkaExactlyOnceDemo.java (hunk 2 above)
Comment (fvaleri) [reply examples/src/main/java/kafka/examples/KafkaExactlyOnceDemo.java:36]: Yep.

File: examples/src/main/java/kafka/examples/KafkaExactlyOnceDemo.java (hunk 3 above)
Comment (fvaleri) [reply examples/src/main/java/kafka/examples/KafkaExactlyOnceDemo.java:69]: Yep. Added in line 35.

File: examples/src/main/java/kafka/examples/KafkaExactlyOnceDemo.java (hunk 4 above)
Comment (fvaleri) [reply examples/src/main/java/kafka/examples/KafkaExactlyOnceDemo.java]: There was no shutdown before the new processor merge. Added now.

**Comment by fvaleri** [comment]
@showuon thanks for keeping the time to review all these PRs. We should be good now.

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the improvement!

**Comment by showuon** [comment]
Failed tests are unrelated:
```
    Build / JDK 8 and Scala 2.12 / org.apache.kafka.connect.integration.ExactlyOnceSourceIntegrationTest.testConnectorBoundary
    Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testTimeouts()
    Build / JDK 17 and Scala 2.13 / kafka.admin.ReassignPartitionsIntegrationTest.testAlterReassignmentThrottle(String).quorum=kraft
    Build / JDK 17 and Scala 2.13 / kafka.api.TransactionsTest.testBumpTransactionalEpoch(String).quorum=kraft
```

**Pull request merged**